Andy/store item page

WishingWell13 opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes #56

# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

Create the store item page at /store/item/[item-uuid] based on these [figma designs](https://www.figma.com/file/cgbeW6vrqgxW0i1vpbUn1C/Store-(Redesign-2023)?type=design&node-id=654-2155&mode=design&t=jCa9Nm6GIE583hHg-0) (check the 'ready for dev' sections).

Requires https://github.com/acmucsd/membership-portal-ui-v2/issues/54.

Encompasses

- Desktop layout for item page

To Do:
- Mobile layout for item page 
- Modal for adding item to cart (Waiting for @SheepTester PR #91)
- Carousel for images

## Changes

- Designed Desktop item page under `src\pages\store\item\[uuid].tsx` 

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [x] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [x] I have run and passed all new and existing Cypress tests. Add screenshots below.
<img width="578" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/2b88ee71-d599-49f2-9272-59958805878d">

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Please include a screenshot of your Cypress testing suite passing successfully.
<img width="576" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/ee0780c2-fe4d-4c47-a059-c01b527b5332">

If you made any visual changes to the website, please include relevant screenshots below.
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/e91805c6-46a9-4299-9998-f264c9226516)

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/f8f041c1-3ed2-4330-b0ca-26211b86bf8f)

<img width="1031" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/15f8158d-a8cf-4d79-befc-9ed2e992d468">

WishingWell13 committed:
Rough outline of SizeSelector

WishingWell13 committed:
Created a file to represent the item page

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Edited config file to be nested under "Store" header

WishingWell13 committed:
Size Toggle V1, Styling Only

Size toggle styling implemented

On selected text goes blue

On hover border goes blue

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
State of SizeToggle now passed up to the item page

Made a temporary edit to the title of the page that displays the current size selected as testing

WishingWell13 committed:
Updated styling of size selector to match Figma

https://www.figma.com/file/cgbeW6vrqgxW0i1vpbUn1C/Store-(Redesign-2023)?type=design&node-id=2028-16321&mode=design&t=nbIfHwdT8aeDdhsd-0

WishingWell13 committed:
Part 1 of AddCartButton styling

Got text color to change depending on if in stock or not (is currently hard coded, not synced to backend)

Button animation still exists, need to make it flat

Text changes based on if in stock or not

Still need to do styling for button if item is out of stock

WishingWell13 committed:
Out of Stock Button and "Please Select Size"

Updated button theming to change based on if in stock or not

A size is now not selected by default, message requesting user to select a size added.

WishingWell13 committed:
Fixed issue with AddCartButton props breaking

From last update, where made it possible for size to be undefined.

WishingWell13 committed:
Began formatting for item header and item page

WishingWell13 committed:
Store Component now linked to API, price added

Price now displays. No discount display yet

TODO: Modify size buttons to be dynamic depending on how many options there are.

WishingWell13 committed:
Size Selection Buttons Now Dynamically Generated

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Buy ability based on lifetime/monthly limits, added item picture

TODO: Might be a good idea to change the "add to cart" message depending on if you have reached lifetime limits

WishingWell13 committed:
Cleaned up some old comments

WishingWell13 committed:
Moved Navbar to top of page

farisashai commented on src/components/common/ThemeToggle/style.module.scss:
```diff
@@ -1,3 +1,7 @@
+@use 'src/styles/vars.scss' as vars;
+// EXAMPLE: vars.$light-primary-1 (see SizeSelector p{} styling)
+// TODO: Update text size, color, background color, border color, etc. to follow vars presets
```
Does this have anything to do with the store item page? Otherwise I would remove and just open a new issue for it

WishingWell13 replied:
Nope, is a marker to help Diamond Staff to get started with this issue: https://github.com/acmucsd/membership-portal-ui-v2/issues/95

Would it help the Diamond Staff / OK to keep it until then and have them delete it?

farisashai replied:
I think it's simpler to track with a Github issue than have to come look at a comment in the repo to remember this

WishingWell13 replied:
Already created an issue here: https://github.com/acmucsd/membership-portal-ui-v2/issues/95

I figured that having it directly in the codebase might give any Diamond Staff who are unfamiliar with the issue system a bridge between "finding the files only through issues" and "having an indicator on the file".

If we still don't like the idea we can delete it.

farisashai replied:
let's delete all of these comments since the feature has been completed already

farisashai commented on src/styles/pages/store/item.module.scss:
```diff
@@ -0,0 +1,18 @@
+.container {
+  display: flex;
+  flex-direction: column;
+  gap: 16px;
```
```suggestion
  gap: 1rem;
```

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

farisashai commented on src/styles/pages/store/item.module.scss:
```diff
@@ -0,0 +1,18 @@
+.navbarBodyDiv {
+  display: flex;
+  flex-direction: column;
+  gap: 32px;
```
```suggestion
  gap: 2rem;
```

farisashai commented on src/styles/pages/store/item.module.scss:
```diff
@@ -0,0 +1,18 @@
+  align-items: center;
+  display: flex;
+  flex-direction: row;
+  gap: 79px;
```
```suggestion
  gap: 5rem;
```

farisashai commented on src/lib/api/StoreAPI.ts:
```diff
@@ -1,14 +1,38 @@
 import { config } from '@/lib';
+import { UUID } from '@/lib/types';
 import type {
```
```suggestion
import type { UUID } from '@/lib/types';
```

farisashai replied:
why is this resolved? it doesn't look like it was changed

WishingWell13 replied:
<img width="767" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/6326e468-c361-4fea-a928-d5dd6a15c1db">

Can you send me a screenshot of your end? Looks good on mine.

SheepTester replied:
I think he just wants you to put `type` between `import` and `{ UUID }` on line 2, because we're only using `UUID` as a type in that file

WishingWell13 committed:
Resolved linting error (unused var)

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+
+  const buttonStyle = inStock ? styles.buttonInStock : styles.buttonNoStock;
+
+  const optionArr: Array<{ value: string; label: string } | '---'> = [];
```
one line conditions like this and the previous variable should just be inlined directly into the html where they're used to reduce clutter

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+  for (let i = 1; i <= maxCanBuy; i += 1) {
+    optionArr.push({ value: `${i}`, label: `${i}` });
+  }
+
```
does this value need to be a string type or can we keep it as a plain number?

WishingWell13 replied:
Yes, it needs to be strings. The `Dropdown` component only takes in strings. 
<img width="124" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/5cc46648-81d7-44c3-a451-bd717001d25f">
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/aa893c7e-869e-4166-9f81-d45a816d57cf)

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+          }}
+          disabled={!(inStock || maxCanBuy > 0)}
+        />
+      </div>
```
```suggestion
          disabled={!inStock || maxCanBuy === 0}
```
I don't think this is logically valid in every situation but someone can correct me if I'm wrong

WishingWell13 replied:
Can you elaborate on what you mean?

farisashai replied:
your boolean condition `!(inStock || maxCanBuy > 0)` expands to `!inStock && !(maxCanBuy > 0)` or `!inStock && maxCanBuy == 0` using demorgans law so it will only disable if the item is not in stock and it cannot be bought but we want it to disable if either of these are true

SheepTester replied:
not sure, but I think he's just not confident his suggestion is 100% correct either

your current expression says that you can add to cart if it's in stock, but if it's not, you can still add to cart if `maxCanBuy > 0`. so currently, you can get a weird situation like this:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/7c0c4ca0-18f9-4fc9-81d0-19901a9595fb)

> You can't buy any more of this item!.

but then it also says

> This item is in stock.

and then it lets you add to cart

farisashai commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,59 @@
+  flex-direction: column;
+  gap: 8px;
+  //   overflow: auto;
+  // padding: 5px 3px;
```
delete comments

farisashai reviewed (changes requested):
haven't looked at a preview but some quick initial nitpicks for code quality

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/lib/api/StoreAPI.ts

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Resolved merge error with calendar

And API type error

WishingWell13 committed:
Minor package changes

alexzhang1618 committed:
Fix package.json

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Make compatible with multiple store gallery images

No carousel yet, uses default photo as photo shown

WishingWell13 committed:
Moved item header to be above sizing buttons

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

farisashai commented on src/styles/pages/store/item.module.scss:
i would rename this file to `src/styles/pages/StoreItemPage.module.scss`. theres no point to nesting the css in folders really at this level

WishingWell13 replied:
Should I also rename the other files in that folder, or leave that for another PR?

farisashai replied:
there's just one file so might as well

farisashai commented on src/styles/pages/store/item.module.scss:
```diff
@@ -0,0 +1,18 @@
+}
+
+.rowContainer {
+  align-items: center;
```
```suggestion
  align-items: flex-start;
```

farisashai commented on src/styles/pages/store/item.module.scss:
```diff
@@ -0,0 +1,18 @@
+  align-items: center;
+  display: flex;
+  flex-direction: row;
+  gap: 5rem;
```
```suggestion
  gap: 5rem;
  justify-content: center;
```

farisashai commented on .vscode/settings.json:
```diff
@@ -3,8 +3,8 @@
-    "source.fixAll": true,
-    "source.organizeImports": true
+    "source.fixAll": "explicit",
+    "source.organizeImports": "explicit"
```
```suggestion
    "source.fixAll": true,
    "source.organizeImports": true
```

farisashai commented on package.json:
```diff
@@ -22,6 +22,9 @@
     "test:ui": "yarn cypress open"
   },
+  "resolutions": {
+    "@types/react": "18.0.21"
```
seeking to learn: what is this for?

WishingWell13 replied:
"Adding resolutions makes sure it uses the correct version of types/react"

See here: https://discord.com/channels/573028991527550986/1187165206305046670/1191216173329694770
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/0bb8a99d-c0ec-4462-939a-ce3e0367e50c)

farisashai commented on src/components/layout/Navbar/index.tsx:
```diff
@@ -125,7 +125,11 @@ const Navbar = ({ user }: NavbarProps) => {
         </Link>
-        <Link className={styles.mobileNavItem} href={config.storeRoute}>
+        <Link
+          onClick={() => setMenuOpen(false)}
```
why is this onClick here for a link?

SheepTester replied:
it's to close the menu after clicking on a menu item for the mobile navbar. this line is no longer needed though

SheepTester commented on .vscode/settings.json:
```diff
@@ -3,8 +3,8 @@
-    "source.fixAll": true,
-    "source.organizeImports": true
+    "source.fixAll": "explicit",
+    "source.organizeImports": "explicit"
```
what does this do

WishingWell13 replied:
It's been automatically switching to that. Unsure why. I must have missed reverting it once. I'll change it back.

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+  const maxCanBuy = Math.min(lifetimeRemaining, montlyRemaining);
+
+  useEffect(() => {
+    // console.log(`currSize ${currSize} inStock ${inStock} inCart ${inCart}`);
```
delete unused comments

a tip for future PRs you open - try to do a full self review before you have anyone else look at it and if you get a comment, see everywhere it applies and fix it so you don't need to read the same comment written multiple times

WishingWell13 replied:
If someone could show me how to get to the view y'all are using to make small commits, that would be helpful for this.

SheepTester replied:
not really a special view or anything, but reviewing all the changes you made https://github.com/acmucsd/membership-portal-ui-v2/pull/94/files can be helpful (albeit it can be a lot to look over), because these are the changes we are reviewing, and it highlights the lines you made. so you could use this to find all the comments you added, and see if they're unused or not

for seeing if a suggestion can be applied elsewhere in your code, you can go back into VS Code and search across all the files (ctrl+shift+F). for example, you could use this to look up some kinds of comments, like `// Justification for disabling rules`, to see if they're necessary or not

for the little suggested commits we make in reviews, we do that by doing

````md
```suggestion
suggested change here
```
````

and then github will show a button to make the commit

SheepTester commented on src/components/layout/Navbar/index.tsx:
```diff
@@ -125,7 +125,11 @@ const Navbar = ({ user }: NavbarProps) => {
         </Link>
-        <Link className={styles.mobileNavItem} href={config.storeRoute}>
+        <Link
+          onClick={() => setMenuOpen(false)}
```
this used to be consistent, but another PR moved this `onClick` handler to the parent element, so for consistency you can now remove this line

WishingWell13 replied:
Removed `onClick` only if I understood that right

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+
+  for (let i = 1; i <= maxCanBuy; i += 1) {
+    optionArr.push({ value: `${i}`, label: `${i}` });
+  }
```
you need to add a check if maxCanBuy === 0 then push { value: '0', label: '0' } so the dropdown renders correctly still

WishingWell13 replied:
Should I also disable the purchase button if the user selects `0` as the amount to buy?

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+      ) : (
+        <p>You can&apos;t buy any more of this item!.</p>
+      )}
+      <p>{`In cart: ${inCart}`}</p>
```
i think this will work but please confirm
```suggestion
      <p>In cart: {inCart}</p>
```

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+          value={inStock ? 'Add to Cart' : 'Out of Stock'}
+          onClick={() => {
+            setInCart(!inCart);
+          }}
```
It looks like this button toggles the item being in the cart or not. You should update the button's label so it says "remove" when it's already in the cart

SheepTester replied:
also, the default border is showing on the button

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/88f856d6-0267-4237-8e72-cd9e26de6084)

we do remove the border by default, but only for `<button>` elements. I think you should change `<input type="button">` to `<button type="button">` for consistency with the rest of the codebase. it would fix the border

WishingWell13 replied:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/f5c39345-b116-4722-b9c0-c3576b371431)
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/54c1c5fd-1f27-4f84-aa9b-c723e28cd165)

The Figma has a border on the button. Should I put the border back?

Without a border it looks like this:
https://github.com/acmucsd/membership-portal-ui-v2/pull/94/files/385c9dfde91b12565c53d13580c374e954e46afd#r1443356378

SheepTester replied:
hmm, I think you should keep the border when the button is greyed out. I think the button looks better without the border when it's blue, but it's up to you whether the border on the blue button looks nicer

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+            value={`${amount}`}
+            onChange={val => setAmount(Number(val))}
+          />
+        </div>
```
the dropdown is empty when `maxCanBuy` is zero; you should just hide it when you can't select the quantity

WishingWell13 replied:
Need a method to test this

Don't see an item that has hit a lifetime limit and is in stock on the testing portal
https://testing.members.acmucsd.com/store

WishingWell13 replied:
This path using the acm@ucsd.edu email: `store/item/18d76c1d-0d77-40a5-81cf-a7438e89a117`

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/a26025c8-c487-4c01-84ed-468f3f03c8ab)

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+  inStock: boolean;
+  inCart: boolean;
+  lifetimeRemaining: number;
+  montlyRemaining: number;
```
typo? probably meant "monthly"

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+
+  if (!mounted) {
+    return null;
+  }
```
Is there a reason why you don't want it to mount on the server side?

WishingWell13 replied:
https://www.npmjs.com/package/next-themes

I thought it would mess up the dark/light mode, but it seems like it won't since we don't use `useTheme()` specifically in this file. Removed it.

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,88 @@
+          height={600}
+        />
+        <div className={styles.container}>
+          <h1>Store Item Page {size}</h1>
```
```suggestion
```
i don't think this is necessary

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,91 @@
+  inCart: boolean;
+  lifetimeRemaining: number;
+  montlyRemaining: number;
+  setInCart: (value: boolean) => void;
```
in React's naming conventions, starting something with `set` should be reserved for something from `useState` within a component. if you want to pass the new `inCart` value back up to the parent component, the prop should start with `on`, like an event handler

```suggestion
  onCartChange: (inCart: boolean) => void;
```

WishingWell13 replied:
<img width="332" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/8c2c0489-8e21-4ae9-a69e-9c919647684d">

farisashai commented on src/components/store/AddCartButton/style.module.scss:
```diff
@@ -0,0 +1,62 @@
+  flex-direction: column;
+  gap: 1rem;
+  overflow: visible;
+  // padding: 5px 3px;
```
delete

farisashai commented on src/components/store/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,37 @@
+  cost: number | undefined;
+  // Justification for disabling rules: This seems to be a false positive.
+  // https://stackoverflow.com/q/63767199/
+  // eslint-disable-next-line no-unused-vars
```
what is this comment for? can you double check your own diff especially when you copy paste code in or delete something that might have been relevant

WishingWell13 replied:
<img width="441" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/046f2228-41e4-4223-b000-eaee46847458">

Originally found in the `Dropdown` component. The placement of the comment was off, fixed in the new update.

Not quite sure what it is. Sent a message in discord.

farisashai commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,57 @@
+
+  .buttonInStock {
+    background-color: var(--theme-primary-2);
+    color: vars.$light-background; // or var(--theme-background);
```
does this work on light and dark mode?

SheepTester replied:
because the button background color is blue, white text color is legible in both dark and light mode. but for consistency with the login page's button, I think it should use `var(--theme-text-on-background1)` or whatever it's called, so the button text is black in dark mode

WishingWell13 replied:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/b005ccd6-99de-41f6-a6e1-72854dc349f7)

Figma says use the `background` color I believe. If there is no functional difference I am good with keeping `var(--theme-text-on-background1)`

SheepTester replied:
oh, I think `var(--theme-background)` might be better. I mixed them up, oops

`var(--theme-text-on-background1)` would be black in light mode, white in dark mode (used for text), while `var(--theme-background)`  is the opposite: white in light mode, black in dark mode (used for the page background color)

the button's blue background is *brighter* in dark mode, so the text color should be darker to compensate, so I think using `var(--theme-background)` for the button text would be better

it's kind of confusing; here's the sign in button for reference:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/93a08080-7689-4266-b7a3-c8e2a26319db) ![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ae249304-d213-45bb-a6f7-99fd386d9335)

farisashai commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,57 @@
+
+  .button {
+    border-color: var(--theme-accent-line-2);
+    border-radius: $button-roundness;
```
variables are unnecessary if they're only used once you can write the value directly in here

WishingWell13 replied:
Which variables are you referring to? I see them be used multiple times:

<img width="127" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/13ee7300-86e8-4470-849c-f7a7221db910">


![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/ea549c71-17d8-41bf-9389-893ab04a87ad)

and headings are standard with the Figma.

WishingWell13 replied:
Saw the other comment, resolved.

SheepTester commented on src/components/store/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,37 @@
+  cost: number | undefined;
+  // Justification for disabling rules: This seems to be a false positive.
+  // https://stackoverflow.com/q/63767199/
+  // eslint-disable-next-line no-unused-vars
```
I don't think you need this here

farisashai commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,56 @@
+export default SizeSelector;
+
+// Links for knowledge:
+// https://medium.com/@vincent.bocquet/typescript-with-react-usestate-hook-f701309384a
```
can you delete this comment everywhere or leave it once at max - it seems like its in a lot of files that have been copy pasted

WishingWell13 committed:
Centered the body of the item page.

SheepTester commented on src/components/store/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,37 @@
+
+  if (!mounted) {
+    return null;
+  }
```
I don't think this is necessary for this component

`useEffect` hurts performance because it makes React render once, run the `useEffect`, then render again. it also prevents the component from being rendered on the server side, which negates the benefits of using NextJS. using `useEffect` is often necessary, but you'd want to avoid it if possible

WishingWell13 replied:
Removed

WishingWell13 committed:
Fixed merge error by updating name of file path

farisashai commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,56 @@
+
+  return (
+    <div className={styles.sizeSelector}>
+      {options.length > 1 && (
```
it's safer to not use conditional rendering like this and replace `condition && <Component />` with `condition ? <Component /> : null` because in the former case, you can get weird rendering bugs where if condition evaluates to 0 or some false values, it will still render that variable's value

farisashai commented on src/components/store/SizeSelector/style.module.scss:
```diff
@@ -0,0 +1,73 @@
+    display: flex;
+    float: left;
+    gap: 12px;
+    // border-style: solid;
```
delete these

farisashai commented on src/components/store/SizeSelector/style.module.scss:
```diff
@@ -0,0 +1,73 @@
+@use 'src/styles/vars.scss' as vars;
+$indicator-size: 30px;
+$switch-gap: 127%;
+$switch-gap-num: calc($switch-gap / 100%);
```
delete any of these variables that you haven't used at least twice

SheepTester commented on src/components/store/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,37 @@
+      <div className={styles.price}>
+        <p>{cost || 'Select Size Option First'}</p>
+        <BlueDiamond />
+      </div>
```
There's a `Diamonds` component that does this already (see the `ItemCard` component to see how it's used), so you don't need blue-diamond.svg

I wonder if it's possible for there to be free items on the store. In that case, it would show "Select Size Option First" instead of "0 diamonds." Maybe you should check to make sure `cost !== undefined` instead of checking if it's truthy, just in case

farisashai reviewed (changes requested):
left comments for everything except the file for the actual store page, i'll take a closer look there when the page is fully done and matches the given design on desktop/mobile (we can skip the image carousel for now and anything that would take too long)

SheepTester commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,56 @@
+  const myOptions = options.map(val => {
+    const tempId = val.metadata?.value ? `${val.metadata.value}${myID}` : `${val.uuid}${myID}`;
+    return (
+      <>
```
Hmm, does React not give a warning about this missing a `key` prop? React usually requires a `key` prop when you return a list of elements, like here, so it can keep track of which is which

You can't add a prop directly to `<></>` though. You'd have to import `Fragment` (which is what `<>` is short for) from React then do

```suggestion
      <Fragment key={val.uuid}>
```

WishingWell13 replied:
<img width="363" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/b1725390-d690-4eac-9516-8d98fa258c81">

WishingWell13 replied:
React was not giving me an error for this, which is odd.

SheepTester replied:
you only need it if you're returning an array of elements. for the `Fragment` in your screenshot, it's not in an array, so you don't need a `key`

WishingWell13 replied:
Which lines were you referring to then? I think I'm misunderstanding.

SheepTester replied:
the lines inside `const myOptions`

SheepTester commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,56 @@
+
+const SizeSelector = ({ currSize, options, setSize }: SizeSelectorProps) => {
+  const [mounted, setMounted] = useState(false);
+  const myID = `${useId()}`;
```
I think `useId` already returns a string so you don't need to cast it into a string again

SheepTester commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,88 @@
+  const currOption =
+    item.options.length <= 1
+      ? item.options[0]
+      : item.options[item.options.findIndex(val => val.metadata?.value === size)];
```
you can also use `.find`

```suggestion
      : item.options.find(val => val.metadata?.value === size);
```

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

SheepTester commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,88 @@
+      <Navbar balance={credits} showBack />
+      <div className={styles.rowContainer}>
+        <Image
+          src={getDefaultMerchItemPhoto(item) || ''}
```
I don't think `getDefaultMerchItemPhoto` returns undefined anymore, so you can get rid of the `|| ''`

```suggestion
          src={getDefaultMerchItemPhoto(item)}
```

SheepTester commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,88 @@
+      <div className={styles.rowContainer}>
+        <Image
+          src={getDefaultMerchItemPhoto(item) || ''}
+          alt="Picture of item "
```
trailing space; you could also make the description a bit more specific, using `item.itemName` or something, like

```suggestion
          alt={`Picture of ${item.itemName}`}
```

SheepTester reviewed (commented):
I know it says "please select a size" like twice, but my eyes still immediately went towards the "Out of stock" button

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/8d941e61-9139-4667-a5fc-e905fb31514a)

I think it should hide things like the add to cart button until you select a size, like

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/abaf0cb7-63c3-4d2d-a200-a2e6fe6153ff)

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update .vscode/settings.json

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/components/store/AddCartButton/index.tsx

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Addressed PR comments, Add to Cart button behavior edit

Updated Add to Cart button to also say "remove from cart"

Quantity Selector now disappear if you can't add to cart or size is not selected.

Add to Cart button disappear if a size is not selected.

Removed various comments

Removed useEffect

WishingWell13 committed:
Updated diamond component in Item Header

Diamond component displays cost

WishingWell13 committed:
Add To Cart button disappears when UserPurchaseLimit reached

WishingWell13 committed:
AddCartButton maxCanBuy saftey checks and styling

Added an option for 0 quantity if user has reached purchase limit

WishingWell13 committed:
Renamed styling files, AddToCart Button Behavior Change

Add To Cart button now displays when user has reached purchase limit, but is greyed out and cannot be clicked

WishingWell13 committed:
Added border back to disabled AddToCart Button

WishingWell13 committed:
Added border back to cart button, cleaned up unused vars

WishingWell13 committed:
Hid toggle when maxCanBuy = 1

Also lifted quantity of items up to uuid.tsx

WishingWell13 committed:
Removed unused comments

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

SheepTester commented on public/assets/icons/blue-diamond.svg:
since you're using the `Diamonds` component now, I don't think you need this file anymore

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,108 @@
+  lifetimeRemaining: number;
+  monthlyRemaining: number;
+  amountToBuy: number;
+  // eslint-disable-next-line no-unused-vars
```
by the way, since I merged my PR #91, I fixed the issue where it thinks the parameter is unused. if you merge in changes from `main`, you can find and remove all `// eslint-disable-next-line no-unused-vars` and it shouldn't give a warning anymore

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,95 @@
+import { Navbar } from '@/components/store';
+import AddCartButton from '@/components/store/AddCartButton';
+import ItemHeader from '@/components/store/ItemHeader';
+import SizeSelector from '@/components/store/SizeSelector';
```
all 3 of these can be imported together in one line from the parent folder with the Navbar

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,108 @@
+              styles.button
+            }`}
+            type="button"
+            title={`${buyButtonText} Button`}
```
I don't think you need the `title` attribute since the button already has a label visible

SheepTester replied:
why is this resolved, the `title` attribute is still here

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,108 @@
+            disabled={!inStock || maxCanBuy === 0}
+          >
+            {' '}
+            {buyButtonText}{' '}
```
out of curiosity, do you need these spaces here? if not, then might as well remove them; `{' '}` is kind of unsightly

```suggestion
            {buyButtonText}
```

SheepTester commented on src/components/store/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,19 @@
+      <h1>{itemName}</h1>
+      <div className={styles.price}>
+        {cost === undefined ? <p>Select Size Option First For Price</p> : <Diamonds count={cost} />}
+      </div>
```
I think you should use sentence case for the no size selected message, "Select size option first for price"

Alternatively, you could just hide the price silently if no size is selected, since it already says "Please select a size" in red below:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/8369b3b1-1e06-4f0b-b8fd-3afe3003dbc3)

SheepTester commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  if (!mounted) {
+    return null;
+  }
```
I don't believe you need this for this component either

SheepTester reviewed (commented):
very good, glad to see the dropdown being hidden when you can only buy one of the item

WishingWell13 committed:
Removed eslint disable lines

WishingWell13 committed:
Removed title attribute from buy button

https://github.com/acmucsd/membership-portal-ui-v2/pull/94#discussion_r1443580814

WishingWell13 committed:
Removed trailing spaces from button text

https://github.com/acmucsd/membership-portal-ui-v2/pull/94#discussion_r1443581247

WishingWell13 committed:
Hide price when size is not selected

Also removed mounted checks from `SizeSelector`

importing only types from UUID

WishingWell13 committed:
Added back other types accidently deleted from last commit

WishingWell13 committed:
Delete blue-diamond.svg

WishingWell13 committed:
Basic mobile formatting implemented

Still need to sync it with Figma

WishingWell13 committed:
Fixed image clipping issue, reduced image gap on mobile

WishingWell13 committed:
Removed vertical scrollbar from mobile view

Used scrollbar hide, may need to look into why is actually overflowing later.

WishingWell13 committed:
Removed unnecessary CSS

SheepTester committed:
Merge remote-tracking branch 'origin/main' into andy/StoreItemPage

SheepTester committed:
Rename store/collections.module.scss to be consistent with other changes in PR

SheepTester committed:
PermissionService.allUserTypes is no longer a function

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,105 @@
+                options={optionArr}
+                value={`${amountToBuy}`}
+                onChange={val => onAmountChange(Number(val))}
+              />
```
Is there a reason why you're using a dropdown instead of a number input for quantity?

SheepTester replied:
currently it crashes the browser and server when the max quantity is large: https://membership-portal-ui-v2-git-andy-storeitempage-acmucsd.vercel.app/store/item/a174f5b2-de6c-44f2-b99c-f36a27221797

SheepTester commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,42 @@
+        </Fragment>
+      ) : null}
+    </div>
+  );
```
What I meant for adding the `key` prop was to add it to elements inside arrays, like `myOptions`, so React can tell the items apart. So below in the suggestion, I added `key` to the `Fragment` inside `map` since it creates an array.

```suggestion
  const myOptions = options.map(val => {
    const tempId = val.metadata?.value ? `${val.metadata.value}${myID}` : `${val.uuid}${myID}`;
    return (
      <Fragment key={tempId}>
        <input
          id={tempId}
          name="state-d"
          type="radio"
          defaultChecked={currSize === val.metadata?.value}
          onClick={() => onSizeChange(val.metadata?.value)}
        />
        <label htmlFor={tempId}>{val.metadata?.value}</label>
      </Fragment>
    );
  });

  return (
    <div className={styles.sizeSelector}>
      {options.length > 1 ? (
        <>
          <h4>Size</h4>
          <form className={styles.switch}>{myOptions}</form>
        </>
      ) : null}
    </div>
  );
```

The fragment at the bottom doesn't need a `key` because React knows that there will it either be one element or none

WishingWell13 committed:
Fixed a linting error with the options dropdown

WishingWell13 committed:
Shrunk some UI items for mobile version of site

WishingWell13 committed:
Set upper limit on quantity of items can buy

WishingWell13 committed:
Deleted unnecessary comments

WishingWell13 committed:
Update src/components/store/SizeSelector/index.tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Fixed various linting errors, removed comments

Also updated gaps

WishingWell13 committed:
Decreased font size of dropdown text

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

alexzhang1618 commented on package.json:
```diff
@@ -23,6 +23,9 @@
     "test:ui": "yarn cypress open"
   },
+  "resolutions": {
+    "@types/react": "18.0.21"
```
Wondering if we still need this since you've switched computers 😁

alexzhang1618 reviewed (changes requested):
Some visual nitpicks:

I don't like how the image can take up variable space on the left column here, I think we should keep the ratio standard and scale the image to fit inside of that.
![Screenshot 2024-01-22 at 10 00 34 PM](https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/fa6cd7c2-f0b9-49c6-96ec-6c29699564cb)
On this one, the variant isn't fitting inside of the variant button
![Screenshot 2024-01-22 at 10 01 02 PM](https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/00024c15-0198-4ae0-8325-9741ee99cb6e)

Additionally, it seems like the variant buttons and some of the fonts/spacings aren't consistent with Figma. For reference: 
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/6942dba9-3676-41d4-8e92-cec2195500ff)
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/e30ba9f0-e2cc-4106-a199-20eb3110ceee)

SheepTester commented on src/components/store/SizeSelector/style.module.scss:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    input:checked {
+      border-color: var(--size-selector-border);
```
this border shouldn't be noticeable since it has zero opacity, right?

SheepTester commented on src/components/store/SizeSelector/style.module.scss:
```diff
@@ -0,0 +1,64 @@
+    font-weight: bold;
+  }
+
+  .switch {
```
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e2920b49-5bdb-4c81-8f5c-a6dd26f040e1)

you should change `width` to `min-width` for the labels (and add some horizontal padding) and add `flex-wrap: wrap;` to `.switch` so if there's not enough room, the options go onto the next line

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/a95b2022-a04f-4176-a793-767dd8efc245)

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,105 @@
+    optionArr.push({ value: `${i}`, label: `${i}` });
+  }
+
+  optionArr.push('----');
```
What is the purpose of the final line?

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/b2c71c52-e9e1-4ce1-8044-0d82457c51e6)

SheepTester commented on src/styles/pages/StoreItemPage.module.scss:
```diff
@@ -0,0 +1,44 @@
+  justify-content: center;
+  width: 95%;
+
+  @media (max-width: 600px) {
```
I think the image can get pretty small before converting into a single column layout. I would increase this `max-width` so the single column layout kicks in earlier

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e2a23669-0194-42c2-9d85-d176711b562c)

This codebase would prefer you use the media breakpoint variables in vars.scss for consistency,  rather than hard-coded values like `600px`. You can search for `@media` in other files to see how they're used

SheepTester commented on src/styles/themes.scss:
```diff
@@ -91,6 +92,7 @@
 
   --theme-mode-indicator: #{vars.$dark-mode-indicator};
   --theme-mode-icon: #{vars.$white};
+  --size-selector-border: #{vars.$dark-primary-4};
```
I think for consistency, since all the other variables start with `--theme-`, this one should too

SheepTester reviewed (changes requested):
Very excellent, just some nitpicks.

This is just a quick design review; I'll do a code review later

- I think you should put "You can buy up to 20 of this item." next to the quantity dropdown
- I think it might not be necessary to say "Item Description." Also, maybe it might look nicer if the description is right under the title/price
- We probably don't have to say "This item is in stock." since the fact that you can press on the "Add to Cart" button says that already

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/0d2abe37-a2d6-4f39-9ce4-4a55f25ac741)

and here, it's a bit redundant saying "Out of stock" or "Limit reached" twice

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/f478dbff-638a-4a1d-a8cc-a5509665402c)

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/4d520b75-8b0c-4a64-91c4-f36fb776921c)

I think just having the disabled button is enough



also, you might want to add `word-break: break-word;` for extreme cases like this:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/d2a3c8ae-121b-4a38-a38d-6f34664d5d60)

https://membership-portal-ui-v2-git-andy-storeitempage-acmucsd.vercel.app/store/item/0702252f-a50b-4948-ad21-e1ca38850413

though admittedly, it probably won't happen in practice

SheepTester commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,93 @@
+            onAmountChange={setAmount}
+          />
+          <h4>Item Description</h4>
+          <p>{item.description}</p>
```
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/de8a8388-47f9-41c3-9cde-782f00a19a31)

You should add `white-space: pre-wrap;` so newline characters are kept. You might also want to increase `line-height` to, say, `1.5`, so the text doesn't look as dense:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/37ccf179-3c64-4d07-82ee-fe22212aae68)

WishingWell13 committed:
New styling changes to adjust spacing

IN PROGRESS

WishingWell13 committed:
Locked width of description and fixed line breaks

Also adjusted title/paragraph sizes and weight

Also centered the webpage

WishingWell13 committed:
Title now properly wraps with very long titles

Also deleted some unused styling classes

Also adjusted one column width threshold.

WishingWell13 committed:
Made padding a little bigger for right column

Also fixed bug where price would cause an extra gap in object header, even when not visible

WishingWell13 committed:
Removed in stock / out of stock messages

Also bumped the "add to cart" button up a little bit

WishingWell13 committed:
@media now uses the media breakpoint variables

WishingWell13 committed:
Renamed variables and adjusted white space behavior

farisashai commented on .vscode/settings.json:
lets try to make sure to revert any changes in the PR's diff that aren't related to the feature you're working on

farisashai commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,99 @@
+            value={buyButtonText}
+            onClick={() => {
+              onCartChange(!inCart);
+            }}
```
```suggestion
            onClick={() => onCartChange(!inCart) }
```

you can inline single line arrow functions

farisashai commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,42 @@
+  gap: 0.9rem;
+
+  h1 {
+    @media (max-width: 600px) {
```
nit: hardcoded media query

farisashai commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,42 @@
+  display: flex;
+  flex-direction: column;
+  flex-wrap: wrap;
+  gap: 0.9rem;
```
super tiny nit: lets use a consistent scale for padding values of values like 0.5rem, 0.75, 1rem, 1.5, 2rem to avoid uneven values everywhere

farisashai commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,41 @@
+    <div className={styles.sizeSelector}>
+      {options.length > 1 ? (
+        <>
+          <h4>Size</h4>
```
i'm not sure if this data comes from the api but i'm pretty sure that store admins can define any subset of item types here - for example, sticker options such as color: red, blue, pink.

if there's a label for these options, we should use that otherwise we need to make this text more generic to encompass any set of variants/options

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,93 @@
+  item: PublicMerchItemWithPurchaseLimits;
+}
+const StoreItemPage = ({ user: { credits }, item }: ItemPageProps) => {
+  const [size, setSize] = useState<string | undefined>(item.options.length <= 1 ? 'Y' : undefined);
```
the variant should default to the first size option so don't have to deal with layout shift of everything moving down when the price becomes visible after selection

farisashai replied:
seeking to learn: what is this default value "Y"

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,93 @@
+const StoreItemPage = ({ user: { credits }, item }: ItemPageProps) => {
+  const [size, setSize] = useState<string | undefined>(item.options.length <= 1 ? 'Y' : undefined);
+  const [inCart, setInCart] = useState<boolean>(false);
+  const [amount, setAmount] = useState<number>(1);
```
seeking to learn: why is this state variable defined in this component if the only place it's ever used is in `<AddCartButton>`

farisashai replied:
same question for the inCart variable

farisashai reviewed (changes requested):
great progress so far - left a couple of nitpicks but two major comments to finalize this feature page

1. I don't think the add to cart button should switch the text to "Remove Cart" after being clicked because this introduces a lot more complicated functionality instead of a single purpose for the button. the remove from cart functionality will be handled on the actual cart page. the button action should always add the selected quantity of the variant to the cart and (in a future PR) should eventually open a modal confirming the item is added to the cart with the two options to "View Cart" or "Continue Shopping" (return to store home page) which match the implemented design
2. you need to find a reasonable max width for the page container and set it so the space between the content doesn't grow unbound on wider screens (you can zoom out a lot to test this) .coverContainer cannot have full width for this, you need to have the container hug the image's square aspect ratio

SheepTester commented on src/components/store/AddCartButton/style.module.scss:
```diff
@@ -44,8 +49,8 @@
-    height: 40px;
-    width: 200px;
+    padding: 0.3rem 1rem;
+    word-break: break-all;
```
it's better to use

```suggestion
    word-break: break-word;
```

`break-all` ignores words and breaks wherever it wants, even when it's not necessary to break a word (fine for Chinese, not great for English), while `break-word` only breaks words if it's absolutely necessary

[here's a demo/documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/word-break)

WishingWell13 committed:
Update src/components/store/AddCartButton/index.tsx

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/components/store/AddCartButton/style.module.scss

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Fixed some gap inconsistencies

WishingWell13 committed:
Update settings.json

WishingWell13 committed:
Update settings.json

WishingWell13 committed:
Made SizeSelector title change based on metadata

WishingWell13 committed:
Fixed bug where add to cart button no longer worked

Was due to comparison for current option not being properly updated with new system (needed a `.value`)

Also made image container always square.

WishingWell13 committed:
Condensed imports in `uuid`, removed "Remove from Cart" text

Note that clicking "Add to Cart" again removes item from cart.

SheepTester commented:
feedback from @farisashai in #124 that better applies to your PR:

> i'm going to leave some comments here that are only relevant for the store item page too and review it all at once
> 
> * when you enter an item page with variants, the first option that isn't out of stock should be selected by default to prevent layout shift from the price appearing after selecting an option

> might be nice to list which collection this item is a part of on the item page if that data exists easily

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+  if (maxCanBuy === 0) {
+    optionArr.push({ value: `0`, label: `0` });
+  }
+
```
This line isn't needed anymore because the dropdown gets hidden when `maxCanBuy` is 0, right?

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+      ) : (
+        <p>You can&apos;t buy any more of this item!.</p>
+      )}
+      {currSize === undefined ? <p className={styles.error}>Please select a size.</p> : null}
```
You should use the variant type thingy instead of "size." For example, for [apple](https://membership-portal-ui-v2-git-andy-storeitempage-acmucsd.vercel.app/store/item/66a2a2e6-464d-4d74-a9f1-08b8ffa4fc8c) it should say "color"

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/5feefbf2-b14b-4e6e-ae8b-308f2b00c9a7)

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+  onAmountChange: (amountToBuy: number) => void;
+}
+
+const AddCartButton = ({
```
I think you could look into a better name for this component since it now has a bit more than just the add to cart button

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+              <h4>Quantity</h4>
+              <Dropdown
+                name={`options${myID}`}
+                ariaLabel={`Dropdown to select the number of items to purchase for ${myID}`}
```
I think you should use the item name and variant rather than `myID`

SheepTester commented on src/components/store/AddCartButton/index.tsx:
```diff
@@ -0,0 +1,93 @@
+            }`}
+            type="button"
+            title={`${buyButtonText} Button`}
+            value={buyButtonText}
```
I don't think you need to set `value` for a `<button>` here, so you should remove it (as well as `title`)

SheepTester commented on src/components/store/AddCartButton/style.module.scss:
```diff
@@ -0,0 +1,70 @@
+      font-style: normal;
+      font-weight: 700;
+      line-height: 1.25rem;
+      line-height: 200%; /* 2.5rem */
```
Is there a reason why you have multiple `font-size`s and `line-height`s here?

```suggestion
      font-size: 1.25rem;
      font-style: normal;
      font-weight: 700;
      line-height: 2.5rem;
```

SheepTester commented on src/components/store/AddCartButton/style.module.scss:
```diff
@@ -0,0 +1,70 @@
+    border: solid;
+    border-color: var(--theme-accent-line-2);
+
+    border-width: 1px;
```
You could use the shorthand here

```suggestion
    border: 1px solid var(--theme-accent-line-2);
```

SheepTester commented on src/components/store/ItemHeader/style.module.scss:
```diff
@@ -0,0 +1,42 @@
+    @media (max-width: vars.$breakpoint-md) {
+      font-size: vars.$card-heading-1;
+    }
+    display: inline;
```
What is this for? I think you can remove `display: inline`

SheepTester commented on src/components/store/SizeSelector/style.module.scss:
```diff
@@ -0,0 +1,61 @@
+    input {
+      opacity: 0;
+      position: absolute;
+    }
```
You should also put `pointer-events: none;` or something because I'm still able to click on the radio for B even though it's on top of A

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/fe00f2e7-f249-454d-b289-bb9fe66aee40)

SheepTester commented on src/styles/pages/StoreItemPage.module.scss:
```diff
@@ -0,0 +1,81 @@
+  flex: none;
+  flex-direction: column;
+  gap: 1.1rem;
+  max-width: 22.5rem;
+  min-width: 18rem;
```
it still feels like the layout depends on the length of the text

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/5ec37fdf-2f82-4a85-a562-850ae88352b1)

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/2a9ff277-c775-4062-8651-e6b33d2f605e)

adding `width: 100%` here fixes this

SheepTester reviewed (approved):
looks good, just a bunch of nit picks

WishingWell13 committed:
Before attempting to remove currOption

I think I can collapse it into selectedOption

WishingWell13 committed:
Made the "please select an option" text to show key value

So instead of always saying "size" or "option" it will show "color" if the options are colors

WishingWell13 committed:
Update src/components/store/AddCartButton/style.module.scss

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Changed AddCartButton to CartOptionsGroup

Also various small bug fixes from PR

- Made the options `input` tags not selectable

- Removed unecessary CSS

WishingWell13 committed:
Merge branch 'andy/StoreItemPage' of https://github.com/acmucsd/membership-portal-ui-v2 into andy/StoreItemPage

WishingWell13 committed:
Removed unnecessary properties, simplified CSS

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,103 @@
+    item.options.find(val => val.metadata?.value === selectedOption?.value) ??
+    item.options.find(option => option.quantity > 0) ??
+    item.options[0] ??
+    null;
```
wondering what the difference is between selectedOption and currItemOption - i'm not sure if two separate variables are needed here. when I suggested the change to use the first item option to set the price field on page load, I was thinking it would be implemented by setting the initial value of selectedOption to `item.options[0]` instead of undefined on line 26

farisashai commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,48 @@
+
+  return (
+    <div className={styles.sizeSelector}>
+      {options.length > 1 ? (
```
if we have access to the options variable in the parent component rendering the `SizeSelector`, can we do this check at that top level instead and say don't render the SizeSelector if this item doesn't support options so we don't have to render an empty div and save it's state

WishingWell13 committed:
Updated code to have one item selected by default

Also no longer have two separate variables for the selectedOption and currItemOption

WishingWell13 committed:
Removed empty div when there are no options

Also fixed a bug where page crashed on no option item pages

WishingWell13 committed:
Removed console.log statement

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Fixed bug where you couldn't purchase no option items

alexzhang1618 commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -0,0 +1,44 @@
+}
+function toTitleCase(str: string) {
+  return str.toLowerCase().replace(/\.\s*([a-z])|^[a-z]/gm, s => s.toUpperCase());
+}
```
For cleanliness, let's move this to utils and add a comment describing it

alexzhang1618 commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,102 @@
+  // item.options.find(val => val.metadata?.value === selectedOption?.value) ??
+  // item.options.find(option => option.quantity > 0) ??
+  // item.options[0] ??
+  // null;
```
Let's clean up these comments if we're not using them

alexzhang1618 reviewed (commented):
For variantless items, should this say Add to Cart instead of Select a Size? Or are we waiting for cart implementation to come through first?
<img width="1440" alt="Screenshot 2024-02-15 at 1 22 40 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/bd5062d1-4068-4f2c-8260-d4647397f3b0">

Besides this, everything else LGTM

WishingWell13 committed:
Button and option name updates

Changed add to cart text to be always white

Fixed bug where optionless items would have "please select a size" as an option.

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Updated components to use Typography, not <p>

Also moved capitalization function to utils and removed comments

WishingWell13 committed:
Merge branch 'andy/StoreItemPage' of https://github.com/acmucsd/membership-portal-ui-v2 into andy/StoreItemPage

alexzhang1618 commented on src/components/store/CartOptionsGroup/index.tsx:
```diff
@@ -0,0 +1,93 @@
+}: CartOptionGroupProps) => {
+  const myID = useId();
+
+  const maxCanBuy = Math.min(20, Math.min(lifetimeRemaining, monthlyRemaining));
```
Let's add a comment explaining the business logic for this?

alexzhang1618 commented on src/components/store/CartOptionsGroup/index.tsx:
```diff
@@ -0,0 +1,93 @@
+
+  for (let i = 1; i <= maxCanBuy; i += 1) {
+    optionArr.push({ value: `${i}`, label: `${i}` });
+  }
```
nit: I think 
```tsx
Array(maxCanBuy).fill().map((_, i) => i+1);
```
would be more efficient? Can add a comment explaining what this is doing also

alexzhang1618 reviewed (approved):
brilliant 🤩 good to merge after these two nits

WishingWell13 committed:
Refactored the toggle value array creation

Also added comment for business logic behind maxCanBuy

farisashai commented:
This should not be a dropdown tbh because there's no reason to list every number when it can get arbitrarily large- it should be a numeric input field that provides an error message if the input is invalid. We should open a separate issue for this fix before merging this since it's not a quick change

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/33165426/7518abb4-0b75-4154-b806-47e7b6c30154)

farisashai commented:
From the perspective of not over complicating components and isolating a single purpose to each one, I think it would make a lot more sense for the Add to Cart button to never change its text. instead, there should be a Boolean prop passed in whether to disable (grey out) the action or not. The error should be explained on the page itself in the spot where it says "You can buy up to 1 of this item" because right now that text is misleading saying you can buy 1 but it's actually out of stock

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/33165426/39fd078b-42eb-4838-9048-0dded22698cd)

farisashai commented:
For whatever reason, it seems like a lot of images that have been uploaded to the store historically have a transparent region in the png. We should account for this by adjusting the design in some way with a border or shadow to highlight the image in separation from the page

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/33165426/3f286340-9cd8-433a-bea2-4975edabdd68)

WishingWell13 committed:
Implemented discount functionality

farisashai commented on src/styles/pages/StoreItemPage.module.scss:
```diff
@@ -0,0 +1,74 @@
+  justify-content: center;
+
+  margin-left: auto;
+  margin-right: auto;
```
shorthand for this is `margin: 0 auto;`

farisashai commented on src/components/store/CartOptionsGroup/index.tsx:
```diff
@@ -0,0 +1,95 @@
+  const optionArr: Array<{ value: string; label: string }> = Array(maxCanBuy)
+    .fill({ value: `${1}`, label: `${1}` })
+    .map((i, index) => {
+      return { value: `${Number(i.value) + index}`, label: `${Number(i.label) + index}` };
```
I don't think you need the fill here - you can just take the original array and map every element to index + 1 since that's already hardcoded anyways. Don't need to store it in a variable

farisashai commented on src/pages/store/item/[uuid].tsx:
```diff
@@ -0,0 +1,108 @@
+
+const StoreItemPage = ({ user: { credits }, item }: ItemPageProps) => {
+  const [selectedOption, setSelectedOption] = useState<PublicMerchItemOption | null | undefined>(
+    item.options.find(option => option.quantity > 0) ?? item.options[0] ?? null
```
just wondering, when would this default to a value that isn't just the first element in the array ?

SheepTester replied:
when `item.options` is empty

our typescript rules make `item.options[0]` a `PublicMerchItemOption | undefined`. but also, backend doesn't enforce that items must have at least one option, so there can be [items with no options](https://membership-portal-ui-v2-git-andy-storeitempage-acmucsd.vercel.app/store/item/13b32fdc-a220-48e4-b432-88aa33019e22)

farisashai commented on src/components/store/CartOptionsGroup/index.tsx:
```diff
@@ -0,0 +1,95 @@
+          )}
+
+          <button
+            className={`${inStock && maxCanBuy > 0 ? styles.buttonInStock : styles.buttonNoStock} ${
```
you used this Boolean condition (or its negation) at least three times in this file from what I've seen. you should probably store it in a Boolean variable isPurchasable and reference that everywhere instead to keep the logic consistent as to exactly what it's checking

farisashai reviewed (commented):
some quick tiny comments but everything seems good now - will approve after we have issues covering all of the earlier comments I mentioned

farisashai commented:
I think we're all aware of this last point but we also need an issue for the add to cart button to open a modal that lets the user confirm adding the item to the actual cart, however Raymond decides to implement that

WishingWell13 committed:
Cleaned up CSS and conditionals by adding variables

farisashai reviewed (approved):
LGTM 🚀
